core/src/main/java/org/infinispan/persistence/spi/TransactionalCacheWriter.java

```diff
@@ -0,0 +1,27 @@
+package org.infinispan.persistence.spi;
+
+import org.infinispan.marshall.core.MarshalledEntry;
+
+import javax.transaction.Transaction;
+import java.util.List;
+
+/**
+ * @author Ryan Emerson
+ */
+public interface TransactionalCacheWriter<K, V> extends AdvancedCacheWriter<K, V> {
+
+   /**
+    * Always write modifications in prepare, as this is the only way we know the FINAL values of the entries. This is
+    * required to handle scenarios where an objects value is changed after the put command has been executed, but before
+    * commit is called on the Tx.
+    * @param transaction
+    * @param modifiedEntries
+    * @param keysToDelete
+    */
+   void prepareWithModifications(Transaction transaction, List<MarshalledEntry> modifiedEntries,
```

Comment:
`throws Throwable` seems a bit broad.  Can we limit this to `PersistenceException`?